File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/java/domain/hierarchy/AbstractTreeNode.java
package com.coupang.springframework.data.requery.java.domain.hierarchy;

import com.coupang.kotlinx.objectx.ToStringBuilder;
import com.coupang.springframework.data.requery.domain.AbstractPersistable;
import io.requery.*;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Set;

/**
 * com.coupang.springframework.data.requery.java.domain.hierarchy.AbstractTreeNode
 *
 * @author debop
 * @since 18. 6. 4
 */
@Getter
@Setter
@Entity
@Table(name = "tree_node")
public class AbstractTreeNode extends AbstractPersistable<Long> {

    private static final long serialVersionUID = -4267441735293906937L;

    @Key
    @Generated
    @Column(name = "nodeId")
    protected Long id;

    @Column(name = "nodeName", length = 48)
    protected String name;

    @ManyToOne
    @ForeignKey(delete = ReferentialAction.SET_NULL, update = ReferentialAction.CASCADE)
    protected AbstractTreeNode parent;

    @OneToMany(mappedBy = "parent", cascade = { CascadeAction.DELETE, CascadeAction.SAVE })
    protected Set<AbstractTreeNode> children;

    @OneToMany(mappedBy = "node")
    protected Set<NodeAttribute> attributes;

    @Embedded
    protected NodePosition nodePosition;

    @Override
    public int hashCode() {
        return Objects.hash(id, name);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name);
    }
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/hierarchy/AbstractTreeNode.kt
package com.coupang.springframework.data.requery.domain.hierarchy

import com.coupang.kotlinx.core.hashOf
import com.coupang.kotlinx.objectx.ToStringBuilder
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*

/**
 * AbstractTreeNode
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Entity
abstract class AbstractTreeNode: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    @get:Column(name = "nodeId")
    abstract override val id: Long?

    @get:Column(name = "nodeName", length = 48)
    abstract var name: String

    @get:ManyToOne
    @get:ForeignKey(delete = ReferentialAction.SET_NULL, update = ReferentialAction.CASCADE)
    abstract var parent: AbstractTreeNode?

    @get:OneToMany(mappedBy = "parent", cascade = [CascadeAction.DELETE, CascadeAction.SAVE])
    abstract val children: MutableSet<AbstractTreeNode>

    @get:OneToMany(mappedBy = "node", cascade = [CascadeAction.DELETE, CascadeAction.SAVE])
    abstract val attributes: MutableSet<AbstractNodeAttribute>

    fun addChild(child: AbstractTreeNode) {
        child.parent = this
        children.add(child)
    }

    fun removeChild(child: AbstractTreeNode) {
        children.remove(child)
        child.parent = null
    }

    fun addAttribute(attr: AbstractNodeAttribute) {
        this.attributes.add(attr)
        attr.node = this
    }

    fun removeAttribute(attr: AbstractNodeAttribute) {
        this.attributes.remove(attr)
        attr.node = null
    }

    override fun hashCode(): Int {
        return hashOf(name)
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("name", name)
    }
}